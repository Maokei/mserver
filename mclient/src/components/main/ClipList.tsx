import * as React from "react";

export const ClipList = () => {
    return <div>ClipList</div>;
};
