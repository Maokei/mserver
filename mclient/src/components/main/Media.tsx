import React from "react";

export interface MediaProps {
    id: number;
    imgSrc: string;
    mediaTitle: string;
    mediaSubtitle: string;
    play: boolean;
    like: boolean;
    playState: boolean;
    likeState: boolean;
    handleToggle: Function;
}

export const Media: React.FC<MediaProps> = ({
    id,
    imgSrc,
    mediaTitle,
    mediaSubtitle,
    playState,
    likeState,
    handleToggle,
}) => {
    const playBtnActive = playState ? "play-btn-active" : "";

    const handleClick = (e: React.MouseEvent<HTMLButtonElement>) => {
        e.preventDefault();
        handleToggle(e.currentTarget.id);
    };

    return (
        <div className="mediaList-item">
            <figure className="image is-48x48">
                <img src={imgSrc} alt={mediaTitle} />
            </figure>
            <p className="title is-4">{mediaTitle}</p>
            <p className="subtitle is-6">{mediaSubtitle}</p>
            <button
                id={id.toString()}
                className={`button is-dark icon like-btn ${
                    likeState ? "like-btn-active" : ""
                }`}
                onClick={handleClick}
            >
                <i className="fas fa-heart"></i>
            </button>
            {!playState ? (
                <button
                    className={`button is-dark icon play-btn ${playBtnActive}`}
                >
                    <i className="fas fa-play"></i>
                </button>
            ) : (
                <button className="button is-dark icon pause-btn">
                    <i className="fas fa-pause"></i>
                </button>
            )}
        </div>
    );
};
