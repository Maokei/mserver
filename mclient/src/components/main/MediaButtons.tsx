import React from "react";

export const MediaButtons = () => {
    return (
        <div data-testid="button-group" className="buttons-wrapper">
            <button className="button is-dark icon icon-previous">
                <i className="fas fa-caret-left"></i>
            </button>
            {/* <button className="button is-dark icon icon-play">
                <i className="fas fa-play"></i>
            </button> */}
            <button className="button is-dark icon icon-pause">
                <i className="fas fa-pause"></i>
            </button>
            <button className="button is-dark icon icon-next">
                <i className="fas fa-caret-right"></i>
            </button>
            <button className="button is-dark icon icon-shuffle">
                <i className="fas fa-random"></i>
            </button>
            <button className="button is-dark icon icon-volume">
                <i className="fas fa-volume-up"></i>
            </button>
        </div>
    );
};
