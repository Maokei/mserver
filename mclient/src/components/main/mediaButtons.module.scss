.wrapper {
	padding-left: 8vw;
	grid-area: button-group;

	display: grid;
	grid-template-columns: repeat(10, 1fr);
	align-self: center;

	.pause,
	.play,
	.previous,
	.next,
	.shuffle,
	.volume {
		justify-self: center;
		align-self: center;
	}

	.play,
	.pause {
		width: 75px;
		height: 75px;
		background-color: var(--color-orange);

		&:hover {
			background-color: lighten(#ea5110, 5%);
		}
	}

	.shuffle {
		grid-column: 9 / span 1;
	}

	@media screen and (max-width: 768px) {
		grid-row: 3 / -1;
		display: flex;
		gap: 10px;
		margin-top: 1em;

		.previous {
			order: 2;
		}
		.play {
			order: 3;
		}
		.next {
			order: 4;
		}
		.shuffle {
			order: 1;
		}
		.volume {
			order: 5;
		}
	}
}
