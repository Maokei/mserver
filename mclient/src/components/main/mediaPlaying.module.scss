.wrapper {
	grid-area: now-playing;

	display: grid;
	grid-template-areas:
		"medium-art ."
		"medium-art medium-album"
		"medium-art medium-title"
		"medium-art medium-artist"
		"medium-art .";

	p {
		text-align: center;
	}

	@media screen and (max-width: 768px) {
		grid-column: 1 / -1;
		grid-row: 2 / 3;
	}
}

.art {
	grid-area: medium-art;
	position: inherit;
	width: 128px;
	height: 128px;
	border: 2px solid;
	border-image-source: linear-gradient(
		141.17deg,
		#41464b 16.39%,
		#1a1b1f 86.27%
	);
	border-radius: 100%;
	box-shadow: 8px 8px 20px 0px rgba(0, 0, 0, 0.3),
		-8px -8px 20px 0px rgba(122, 134, 151, 0.15);
}

.album {
	grid-area: medium-album;
}

.title {
	grid-area: medium-title;
	color: var(--color-orange);
	filter: blur(0.75px);
}

.artist {
	grid-area: medium-artist;
	color: var(--color-grey);
}
