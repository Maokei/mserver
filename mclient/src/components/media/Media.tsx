import { useState, useEffect, useRef } from "react";
import styles from "./media.module.scss";
import Player from "./player/Player";

const Media = ({
	baseUrl,
	id,
	title,
	foreignId,
}: {
	baseUrl: string;
	id: string;
	title: string;
	foreignId: string;
}) => {
	const [clips, setClips] = useState([]);
	const [loading, setLoading] = useState(true);
	const [error, setError] = useState(null);
	// const [clips, setClips] = useState(dummyData);
	const [isPlaying, setIsPlaying] = useState(false);
	const [currentMedia, setCurrentMedia] = useState({});
	console.log(currentMedia);

	const mediaElement = useRef<any>();

	useEffect(() => {
		const getData = async () => {
			try {
				const response = await fetch(
					"http://localhost:8080/api/v1/media"
				);
				if (!response.ok) {
					throw new Error(
						`This is a HTTP error with a status of ${response.status}`
					);
				}
				const data = await response.json();
				setClips(data);
				setError(null);
			} catch (error: any) {
				setError(error.message);
				setClips([]);
			} finally {
				setLoading(false);
			}
		};
		getData();
	}, []);

	useEffect(() => {
		if (isPlaying) {
			mediaElement.current?.play();
		} else {
			mediaElement?.current?.pause();
		}
	}, [isPlaying]);

	const onPlaying = () => {
		const duration = mediaElement.current?.duration;
		const currentTime = mediaElement.current?.currentTime;

		setCurrentMedia({
			// ...currentMedia,
			// @ts-ignore
			progress: (currentTime / duration) * 100,
			length: duration,
		});
	};

	return (
		<>
			<li key={id} className={`${styles.item} card`}>
				<div className="card-content">
					{/* {currentMedia && !currentMedia.mediaUrl.includes("mp4") ? ( */}
					<div className={`card-image ${styles.imageContainer}`}>
						<figure className="image is-4by3">
							<img
								src={
									"https://bulma.io/images/placeholders/1280x960.png"
								}
								alt={title}
							/>
						</figure>
					</div>
					{/* ) : null} */}
					<div className={`${styles.contentContainer} content`}>
						{/* {currentMedia.mediaUrl.includes("mp4") ? ( */}
						<video id="media-file">
							<source src={`${baseUrl}/${foreignId}`} />
							Your browser does not support the <code>
								audio
							</code>{" "}
							element.
						</video>
						{/* <video
						src={currentMedia.mediaUrl}
						ref={mediaElement}
						onTimeUpdate={onPlaying}
						>
							<source
								src={`${baseUrl}/${foreignId}`}
								onTimeUpdate={onPlaying}
							/>
							Your browser does not support the <code>audio</code>{" "}
							element.
						</video> */}
						) : (
						<audio
						// src={currentMedia.mediaUrl}
						// ref={mediaElement}
						// onTimeUpdate={onPlaying}
						>
							<source src={`${baseUrl}/${foreignId}`} />
							Your browser does not support the <code>
								audio
							</code>{" "}
							element.
						</audio>
						{/* )} */}
						<div className="controls">
							{currentMedia && (
								<Player
									mediaElement={mediaElement}
									clips={clips}
									setClips={setClips}
									isPlaying={isPlaying}
									setIsPlaying={setIsPlaying}
									// @ts-ignore
									currentMedia={currentMedia}
									setCurrentMedia={setCurrentMedia}
								/>
							)}
						</div>
					</div>
				</div>
			</li>
		</>
	);
};

export default Media;
