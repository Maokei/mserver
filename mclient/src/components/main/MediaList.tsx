import * as React from "react";
import { Media, MediaProps } from "./Media";

interface MediaListProps {
    items: MediaProps[];
}

export const MediaList: React.FC<MediaListProps> = ({ items }) => {
    return (
        <section data-testid="mediaList-wrapper" className="mediaList-wrapper">
            <div className="divider"></div>
            <div data-testid="mediaList" className="mediaList clips">
                {items.map((item) => (
                    <Media
                        key={item.id}
                        id={item.id}
                        imgSrc={item.imgSrc}
                        mediaTitle={item.mediaTitle}
                        mediaSubtitle={item.mediaSubtitle}
                        like={item.like}
                        play={item.play}
                        // onPlayClick={handlePlayClick}
                    />
                ))}
            </div>
        </section>
    );
};
