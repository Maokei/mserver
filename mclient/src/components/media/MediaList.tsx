import React, { useEffect, useState } from "react";
import Media from "./Media";
import styles from "./mediaList.module.scss";

const MediaList = () => {
    const [data, setData] = useState([]);
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState(null);

    useEffect(() => {
        const getData = async () => {
            try {
                const response = await fetch(
                    "http://localhost:8080/api/v1/media"
                );
                if (!response.ok) {
                    throw new Error(
                        `This is a HTTP error with a status of ${response.status}`
                    );
                }
                const data = await response.json();
                setData(data);
                setError(null);
            } catch (error: any) {
                setError(error.message);
                setData([]);
            } finally {
                setLoading(false);
            }
        };
        getData();
    }, []);

    return (
        <main className={styles.container}>
            <h3 className="is-size-3 has-text-white mt-6">MediaPlayer</h3>
            <>
                {loading && <div>A moment please...</div>}
                {error && (
                    <div>{`There is a problem fetching media data - ${error}`}</div>
                )}
                <ul className={styles.list}>
                    {data &&
                        data.map(
                            ({
                                id,
                                title,
                                foreignId,
                            }: {
                                id: string;
                                title: string;
                                foreignId: string;
                            }) => (
                                <li key={id} className={styles.item}>
                                    <figure className="image is-48x48">
                                        <img
                                            className="is-rounded"
                                            src={
                                                "https://bulma.io/images/placeholders/1280x960.png"
                                            }
                                            alt={title}
                                        />
                                    </figure>
                                    <h3 className="title is-3 is-size-6-mobile">
                                        {title}
                                    </h3>
                                    <video
                                        src={`http://localhost:8080/api/v1/media/${foreignId}`}
                                        controls
                                        preload="none"
                                    ></video>
                                </li>
                            )
                        )}
                </ul>
            </>
        </main>
    );
};

export default MediaList;
