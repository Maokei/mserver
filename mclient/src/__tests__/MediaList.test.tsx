import * as ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import { getByTestId } from "@testing-library/react";
import { shallow } from "enzyme";

import { MediaList } from "../components/main/MediaList";
import { Media } from "../components/main/Media";
import data from "../dummyData.json";

let container: HTMLDivElement;

beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
    ReactDOM.render(
        <MediaList items={data} handleToggleLike={() => console.log("test")} />,
        container
    );
});

afterEach(() => {
    document.body.removeChild(container);
    container.remove();
});

describe("Test MediaList component", () => {
    it.skip("should render component", () => {
        const tree = renderer
            .create(
                <MediaList
                    items={data}
                    handleToggleLike={() => console.log("test")}
                />
            )
            .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it("should find MediaList", () => {
        expect(getByTestId(container, "mediaList-wrapper")).toBeTruthy();
    });

    it("should have children", () => {
        const wrapper = shallow(
            <MediaList
                items={data}
                handleToggleLike={() => console.log("test")}
            />
        );

        expect(
            wrapper.containsMatchingElement(
                <Media
                    key={0}
                    songItem={data[0]}
                    handleToggleLike={() => console.log("test")}
                />
            )
        ).toEqual(false);
    });
});
