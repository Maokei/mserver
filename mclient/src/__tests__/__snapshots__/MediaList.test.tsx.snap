// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test MediaList component should render component 1`] = `
<section
  className="mediaList-wrapper"
  data-testid="mediaList-wrapper"
>
  <div
    className="divider"
  />
  <div
    className="mediaList clips"
    data-testid="mediaList"
  >
    <div
      className="mediaList-item"
    >
      <figure
        className="image is-48x48"
      >
        <img
          alt="Dark Imperium ch1"
          src="https://bulma.io/images/placeholders/96x96.png"
        />
      </figure>
      <p
        className="title is-4"
      >
        Dark Imperium ch1
      </p>
      <p
        className="subtitle is-6"
      >
        Guy Haley
      </p>
      <button
        className="button is-dark icon like-btn "
      >
        <i
          className="fas fa-heart"
        />
      </button>
      <button
        className="button is-dark icon play-btn play-btn-active"
      >
        <i
          className="fas fa-play"
        />
      </button>
    </div>
    <div
      className="mediaList-item"
    >
      <figure
        className="image is-48x48"
      >
        <img
          alt="Dark Imperium ch2"
          src="https://bulma.io/images/placeholders/96x96.png"
        />
      </figure>
      <p
        className="title is-4"
      >
        Dark Imperium ch2
      </p>
      <p
        className="subtitle is-6"
      >
        Guy Haley
      </p>
      <button
        className="button is-dark icon like-btn "
      >
        <i
          className="fas fa-heart"
        />
      </button>
      <button
        className="button is-dark icon play-btn play-btn-active"
      >
        <i
          className="fas fa-play"
        />
      </button>
    </div>
    <div
      className="mediaList-item"
    >
      <figure
        className="image is-48x48"
      >
        <img
          alt="Dark Imperium ch3"
          src="https://bulma.io/images/placeholders/96x96.png"
        />
      </figure>
      <p
        className="title is-4"
      >
        Dark Imperium ch3
      </p>
      <p
        className="subtitle is-6"
      >
        Guy Haley
      </p>
      <button
        className="button is-dark icon like-btn "
      >
        <i
          className="fas fa-heart"
        />
      </button>
      <button
        className="button is-dark icon play-btn play-btn-active"
      >
        <i
          className="fas fa-play"
        />
      </button>
    </div>
    <div
      className="mediaList-item"
    >
      <figure
        className="image is-48x48"
      >
        <img
          alt="Dark Imperium ch4"
          src="https://bulma.io/images/placeholders/96x96.png"
        />
      </figure>
      <p
        className="title is-4"
      >
        Dark Imperium ch4
      </p>
      <p
        className="subtitle is-6"
      >
        Guy Haley
      </p>
      <button
        className="button is-dark icon like-btn "
      >
        <i
          className="fas fa-heart"
        />
      </button>
      <button
        className="button is-dark icon play-btn play-btn-active"
      >
        <i
          className="fas fa-play"
        />
      </button>
    </div>
    <div
      className="mediaList-item"
    >
      <figure
        className="image is-48x48"
      >
        <img
          alt="Dark Imperium ch5"
          src="https://bulma.io/images/placeholders/96x96.png"
        />
      </figure>
      <p
        className="title is-4"
      >
        Dark Imperium ch5
      </p>
      <p
        className="subtitle is-6"
      >
        Guy Haley
      </p>
      <button
        className="button is-dark icon like-btn "
      >
        <i
          className="fas fa-heart"
        />
      </button>
      <button
        className="button is-dark icon play-btn play-btn-active"
      >
        <i
          className="fas fa-play"
        />
      </button>
    </div>
  </div>
</section>
`;
